File: MyPlannerKT/src/main/kotlin/ro/bogdannegoita/myplannerkt/domain/PlanInvite.kt
package ro.bogdannegoita.myplannerkt.domain

import ro.bogdannegoita.myplannerkt.commons.dtos.ApplicationUserDto
import ro.bogdannegoita.myplannerkt.commons.types.InviteStatus
import ro.bogdannegoita.myplannerkt.commons.types.InviteStatus.ACCEPTED
import ro.bogdannegoita.myplannerkt.commons.types.InviteStatus.DECLINED
import ro.bogdannegoita.myplannerkt.commons.dtos.PlanInviteDto
import ro.bogdannegoita.myplannerkt.domain.factories.DomainProvider
import ro.bogdannegoita.myplannerkt.persistence.daos.PlanInviteDao
import java.time.LocalDateTime
import java.util.*

class PlanInvite(
    private val data: PlanInviteDto,
    private val dao: PlanInviteDao,
    private val domainProvider: DomainProvider,
) : Comparable<PlanInvite> {
    val id: UUID = data.id
    var status by data::status
        private set
    val createdAt by data::createdAt
    var respondedAt by data::respondedAt
        private set

    val plan: Lazy<Plan> = lazy { domainProvider.plan(dao.getPlan(id)) }
    val sender: Lazy<ApplicationUserDto> = lazy { dao.getSender(id) }
    val recipient: Lazy<ApplicationUserDto> = lazy { dao.getRecipient(id) }

    fun accept() {
        updateStatus(ACCEPTED)
    }

    fun decline() {
        updateStatus(DECLINED)
    }

    private fun updateStatus(value: InviteStatus) {
        status = value
        respondedAt = LocalDateTime.now()
        dao.updateStatus(id, status, respondedAt!!)
    }

    override fun compareTo(other: PlanInvite) = createdAt.compareTo(other.createdAt)
}

File: MyPlannerWeb/src/models/PlanInvite.js
import ApplicationUser from '@/models/ApplicationUser.js';
import Plan from '@/models/Plan.js';
import InviteStatus from '@/models/types/InviteStatus.js';

export default class PlanInvite {
	id;
	status;
	createdAt;
	respondedAt;
	plan;
	sender;
	recipient;

	constructor(response) {
		if (!response) return;
		this.id = response.id;
		this.status = response.status;
		this.createdAt = response.createdAt;
		this.respondedAt = response.respondedAt;
		this.plan = new Plan(response.plan);
		this.sender = new ApplicationUser(response.sender);
		this.recipient = new ApplicationUser(response.recipient);
	}

	get isPending() {
		return this.status === InviteStatus.PENDING;
	}
}
